### PR TITLE
ref(notifications): Consolidate ActivityNotifications and RuleNotifications

### PR DESCRIPTION
**Pull request by mgaeta**
We're trying to move a lot of Notifications responsibility out of the "sentry.mail" module and into the "sentry.notifications" module. This is going to be a pretty large diff so we're splitting the work into ~200 line chunks.

This chunk replaces the `rule_notify` code path with a new class that extends `BaseNotification`.

### CODE REVIEW

File: src/sentry/mail/notifications.py [hunk 1]
```diff
@@ -59,7 +65,17 @@ def can_users_unsubscribe(notification: BaseNotification) -> bool:
 
 
 def log_message(notification: BaseNotification, user: User) -> None:
-    pass
+    if isinstance(notification, AlertRuleNotification):
```
Comment (ceorourke) [inline src/sentry/mail/notifications.py:68]: should there be a log for ActivityNotifications?

File: src/sentry/notifications/activity/base.py [hunk 2]
```diff
@@ -17,7 +18,7 @@
 from sentry.types.integrations import ExternalProviders
 
 
-class ActivityNotification(BaseNotification):
+class ActivityNotification(BaseNotification, ABC):
```
Comment (ceorourke) [inline src/sentry/notifications/activity/base.py:21]: what is ABC and why is it needed here?

Comment (mgaeta) [reply src/sentry/notifications/activity/base.py:21]: [ABC](https://docs.python.org/3/library/abc.html) is python's way of marking a class as abstract. That means that 
```
notification = ActivityNotification(...)
```
is illegal, but you can still instantiate objects for classes that inherit from it.

File: src/sentry/mail/notifications.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/mail/notifications.py:68]: We aren't currently logging for ActivityNotifications but I'll add it in.

Comment (ceorourke) [reply src/sentry/mail/notifications.py:68]: will it be added in this PR or later on?

Comment (mgaeta) [reply src/sentry/mail/notifications.py:73]: Just added `ActivityNotification`.

**Review by ceorourke — APPROVED** [review]

**Pull request merged**